lib/Core/BuildEngine.cpp

```diff
@@ -1534,6 +1532,23 @@ class BuildEngineImpl : public BuildDBDelegate {
     // and \see RuleInfo::isComplete().
     ++currentEpoch;
 
+    if (!traceFile.empty()) {
+      auto trace = llvm::make_unique<BuildEngineTrace>();
+
+      std::string error;
+      if (!trace->open(traceFile, &error))
+        delegate.error(error);
+
+      this->trace = std::move(trace);
+    }
+
+    llbuild_defer {
+      std::string error;
+      if (!trace->close(&error)) {
+        delegate.error(error);
+      }
+    };
```

Comment:
Should we be clearing out/storing nullptr to `this->trace` here too?

Reply:
Good point; done!